_posts/2019-06-04-flink-network-stack.md

```diff
@@ -0,0 +1,324 @@
+---
+layout: post
+title: "A Deep-Dive into Flink's Network Stack"
+date: 2019-06-04T07:00:00.000Z
+authors:
+- Nico:
+  name: "Nico Kruber"
+  
+
+excerpt: Flink’s network stack is one of the core components that make up Apache Flink's runtime module sitting at the core of every Flink job. In this post, we look at the abstractions exposed to the stream operators and detail their physical implementation and various optimisations in Apache Flink. 
+---
+
+Flink’s network stack is one of the core components that make up the `flink-runtime` module and sit at the heart of every Flink job. It connects individual work units (subtasks) from all TaskManagers. This is where your streamed-in data flows through and it is therefore crucial to the performance of your Flink job for both the throughput as well as latency you observe. In contrast to the coordination channels between TaskManagers and JobManagers which are using RPCs via Akka, the network stack between TaskManagers relies on a much lower-level API using Netty.
+
+Before digging any deeper, we will first have a high-level look at what abstractions are exposed to the stream operators and then go into detail on the physical implementation and various optimisations Flink did. We will briefly present the result of these optimisations and Flink’s trade-off between throughput and latency. A future blog post will elaborate more on the tuning parameters and anti-patterns to avoid.
+
+# Logical View
+
+Flink’s network stack provides the following logical view to the subtasks when communicating with each other, for example during a network shuffle as required by a `keyBy()`.
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/2019-06-04-network-stack/flink-network-stack1.png" width="400px" alt="Logical View on Flink's Network Stack"/>
+</center>
+<br>
+
+It abstracts over the different settings of the following three concepts:
+
+* Subtask output type (`ResultPartitionType`):
+    * **pipelined (bounded or unbounded):**
+    Sending data downstream as soon as it is produced, potentially one-by-one, either as a bounded or unbounded stream of records.
+    * **blocking:**
+    Sending data downstream only when the full result was produced.
+
+* Scheduling type:
+    * **all at once (eager):**
+    Deploy all subtasks of the job at the same time (for streaming applications).
+    * **next stage on first output (lazy):**
+    Deploy downstream tasks as soon as any of their producers generated output.
+    * **next stage on complete output:**
+    Deploy downstream tasks when any or all of their producers have generated their full output set.
+
+* Transport:
+    * **high throughput:**
+    Sending each record one-by-one obviously does not give the best throughput. For this, Flink has to buffer a bunch of records into its network buffers and send them altogether.
```

Comment:
```suggestion
    Instead of sending each record one-by-one, Flink buffers a bunch of records into its network buffers and send them altogether. This reduces the cost per record and leads to high throughput.
```

Reply:
Just a suggestion :)